registry/src/metadata/memory.rs

```diff
@@ -0,0 +1,95 @@
+use super::*;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::Mutex;
+use tonic::async_trait;
+
+type MemoryMetadataMap = Arc<Mutex<HashMap<String, Mutex<HashMap<String, PackageManifest>>>>>;
+
+/// InMemory provider for MetadataStorage
+#[derive(Debug)]
+pub struct InMemoryMetadataStorage {
+    packages: MemoryMetadataMap,
+}
+
+impl Default for InMemoryMetadataStorage {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl InMemoryMetadataStorage {
+    /// Creates a new InMemoryMetadataStorage
+    pub fn new() -> Self {
+        InMemoryMetadataStorage {
+            packages: Arc::new(Mutex::new(HashMap::default())),
+        }
+    }
+}
+
+#[async_trait]
+impl MetadataStorage for InMemoryMetadataStorage {
+    /// fetches the version from the storage
+    async fn get_version(
+        &self,
+        package: PackageVersion,
+    ) -> Result<PackageManifest, MetadataStorageError> {
+        let Ok(packages) = self.packages.lock() else {
+            return Err(MetadataStorageError::InternalStuff);
+        };
+
+        let name_string = package.package.to_string();
+        let version_string = package.version.to_string();
+
+        let Some(versions_mutex) = packages.get(name_string.as_str()) else {
+            return Err(MetadataStorageError::PackageMissing(
+                name_string,
+                version_string,
+            ));
+        };
+
+        let versions = versions_mutex.lock().unwrap();
```

Comment:
Unwrapping is never a good idea!

So you could either propagate the `Err` or you can use tokios mutex impl, which lets you await the lock (im 99% confident that you don't need to handle errors in that case)